### PR TITLE
Feat/improved weighting

### PR DESCRIPTION
**Pull request by sondrenaerland**

### CODE REVIEW

File: model/construction_model.py [hunk 1]
```diff
@@ -2,15 +2,12 @@
...
 from model.base_variables import BaseVariables
 from model.construction_objective import ConstructionObjective
-from utils import weights
 
 
 class ConstructionModel(BaseModel):
     def __init__(self, name, problem, data):
         super(ConstructionModel, self).__init__(name, problem, data, mip_focus=1, solution_limit=1)
 
-        self.weights = weights.get_weights()
```
Comment (ekallevik) [inline model/construction_model.py:12]: Is this intentional? From what I can see `self.weigths` is used on line 44.

Comment (ekallevik) [reply model/construction_model.py:12]: I would suggest removing the parameter from this class altogether, and rather just initialize this variable directly in `ConstructionObjective`

File: model/optimality_model.py [hunk 2]
```diff
@@ -2,15 +2,13 @@
...
 from model.optimality_variables import OptimalityVariables
-from utils.weights import get_weights
+
 
 
 class OptimalityModel(BaseModel):
     def __init__(self, name, problem, data):
         super(OptimalityModel, self).__init__(name, problem, data)
 
-        self.weights = get_weights()
```
Comment (ekallevik) [inline model/optimality_model.py:12]: Seems to still be in use on line 36

File: model/optimality_objective.py [hunk 3]
```diff
@@ -65,9 +65,11 @@ def add_objective_for_optimal_solution(self, weights, f, g, delta):
         self.model.setObjective(
             quicksum(f["plus"][e] - f["minus"][e] for e in self.employees)
             + weights["lowest fairness score"] * (g["plus"] - g["minus"])
-            - weights["demand_deviation"]
+            - weights["demand deviation"]
```
Comment (ekallevik) [inline model/optimality_objective.py:68]: Immediately I do not see the need for three different weights for demand deviation. I would think just one weight for positive deviation might suffice if the negative deviation is used as a baseline. What is your intention on using three weights?

File: utils/weights.py [hunk 4]
```diff
@@ -1,17 +1,174 @@
...
-        "contracted hours": 1,
-        "partial weekends": 1,
-        "isolated working days": 1,
-        "isolated off days": 1,
-        "consecutive days": 1,
-        "backward rotation": 1,
-        "preferences": 1,
-        "lowest fairness score": 1,
         "demand_deviation": 5,
+        "shift_duration": 15,
```
Comment (ekallevik) [inline utils/weights.py:39]: Is this actually a weight?

File: utils/weights.py [hunk 5]
```diff
@@ -1,17 +1,174 @@
...
+    """
+    Weekly rest contributes positively to the fairness score (if additional hours are granted), different from the
+    other fairness aspects. Thus it is necessary to correct the weight, so that it actually represents the intention
+    defined in set_weights.
+
+    NOTE:   Preferences are also contributing positively to the fairness score, but these respective weights are
+            corrected naturally in the generate_preferences-function.
+    """
+
+    weights["rest"] = 1 / weights["rest"]
```
Comment (ekallevik) [inline utils/weights.py:74]: Why not just define it as `rest = 1/20` in `set_weights` if one additional hour of rest is equal to 1/20th of demand? 

I think it will make it easier to understand if all weights are measured against the common baseline and not inverted after initialization.

File: utils/weights.py [hunk 6]
```diff
@@ -1,17 +1,174 @@
...
+
+    return weights
+
+
+def add_demand_deviation_weight(weights):
+    """
+    Weight of demand deviation is not an input value and is thus treated in a separate function. The weight is set
+    to 1, serving as the reference value for remaining weights. This value should not be changed.
```
Comment (ekallevik) [inline utils/weights.py:82]: The simplest and most robust way to achieve this is to not set the weight explicitly. The absence of a value implies a weight of 1

Comment (ekallevik) [reply utils/weights.py:82]: Disregard this. I now see that the intent is to handle scaling based on time steps.

File: utils/weights.py [hunk 7]
```diff
@@ -1,17 +1,174 @@
...
+    """
+
+    weights["demand deviation"] = 1
+
+    return weights
+
+
+def scale_weights_to_hours(weights, time_set):
+    """
+    Demand deviation, contracted hours and preferences needs to be scaled to hours, as this is the reference time step.
```
Comment (ekallevik) [inline utils/weights.py:92]: Wouldn't this convert _from_ hours?

File: utils/weights.py [hunk 8]
```diff
@@ -1,17 +1,174 @@
...
+
+
+def scale_contracted_hours_relatively(weights, employee_contracted_hours):
+    """
+    Contracted hours weight is scaled for each employee based on the ratio between contracted hours and
+    default contracted hours. Scaling factor allows for
+    """
+
+    contracted_hours_weights = tupledict()
+    scaling_factor = 1.0
```
Comment (ekallevik) [inline utils/weights.py:123]: Why is this needed? Doubling this is equivalent to just doubling the base contracted hours weight if I am not mistaken

File: utils/weights.py [hunk 9]
```diff
@@ -1,17 +1,174 @@
...
+
+    for e in employee_contracted_hours:
+        contracted_hours_weights[e] = scaling_factor * weights["contracted hours"] * \
+                                      (DEFAULT_CONTRACTED_HOURS / employee_contracted_hours[e])
+
+    weights["contracted hours"] = contracted_hours_weights
```
Comment (ekallevik) [inline utils/weights.py:129]: I think you can make your code cleaner and more Pythonic by returning only the relevant weights from these three scaling functions and updating `scale_weights_relatively()` 108-110 to: 

```
weights["contracted hours"] = scale_contracted_hours_relatively(weights, employee_contracted_hours)
weights["lowest fairness score"] = scale_least_favoured_employee(weights, num_employees)
weights["demand deviation"] = scale_demand_deviation(weights, num_employees)
```

File: utils/weights.py [hunk 10]
```diff
@@ -1,17 +1,174 @@
...
+
+    return weights
+
+
+def scale_least_favoured_employee(weights, num_employees):
```
Comment (ekallevik) [inline utils/weights.py:134]: Just a small tip: Don't Repeat Yourself is one of the core principles of programming. If you want to improve the quality of your code I would suggest (in addition to previous suggestion) changing line 110-110 to 
```
weights["lowest fairness score"] = scale_against_staff_size(weights["lowest fairness score"] , num_employees)
weights["demand deviation"] = scale_against_staff_size(weights["demand deviation"] , num_employees)
``` 

You can then replace `scale_least_favoured_employee` and `scale_demand_deviation` with 
```
def scale_against_staff_size(weight, num_employees):
    
    return weight * num_employees

```

File: model/construction_model.py (hunk 1 above)
Comment (sondrenaerland) [reply model/construction_model.py:12]: Weights are initialized in base_model instead, so the intention is that both construction_model and optimality_model can use self.weights, as they both are built on the base_model. Correct e if this i wrong!

File: model/optimality_model.py (hunk 2 above)
Comment (sondrenaerland) [reply model/optimality_model.py:12]: See the same argumentation in the previous review.

File: model/construction_model.py (hunk 1 above)
Comment (ekallevik) [reply model/construction_model.py:12]: Did not notice that, but that is a really good solution! Resolving this comment.

File: model/optimality_objective.py (hunk 3 above)
Comment (sondrenaerland) [reply model/optimality_objective.py:68]: Good point! The standard "demand deviation" could be removed completely! I still want to keep the "excess demand deviation factor" and "deficit demand deviation factor", as i might be desirable to down or up-scale both of these weights from the reference value 1.

Comment (ekallevik) [reply model/optimality_objective.py:68]: Yeah, you have to at least have two weights here. If not it would be impossible to scale according to `time_step` and `num_employees`

File: utils/weights.py (hunk 4 above)
Comment (sondrenaerland) [reply utils/weights.py:39]: Bad naming. Should represent the weighting of shifts that are either longer or shorter than the desired shift duration. Do you have a possible better name for it?

Comment (ekallevik) [reply utils/weights.py:39]: What about splitting it into `short_shift_duration` and `long_shift_duration`?

File: utils/weights.py (hunk 5 above)
Comment (sondrenaerland) [reply utils/weights.py:74]: I believe it is easier to understand (when setting weights) that setting the weight to 20 means that you are willing to sacrifice 20 hours of demand deviation to get 1 additional hour of weekly rest. This is more intuitively than setting 1/20. This is not in line with how the remaining weights are set.

File: utils/weights.py (hunk 4 above)
Comment (ekallevik) [reply utils/weights.py:39]: Somewhat related: how do you use this weight? Is there a penalty for each time a `short_shift` is used?  Or for each created `short_shift`?

File: utils/weights.py (hunk 7 above)
Comment (sondrenaerland) [reply utils/weights.py:92]: What do you mean? The point is: If weight is set to 4 for demand deviation per hour (bad example, but the most intuitively) and the time step is 0.25, the true weight of demand deviation per hour would be 4 * 4, since every time period of 0.25 hour would get the weight 4. Therefore, the weight is scaled based on the time step - in this case reduced to 1.

File: utils/weights.py (hunk 8 above)
Comment (sondrenaerland) [reply utils/weights.py:123]: Experimented with different scaling of the weights, looked to change how the optimality model behaved. Included it, as it might be worth testing on a later stage.

File: utils/weights.py (hunk 4 above)
Comment (sondrenaerland) [reply utils/weights.py:39]: This is used to weight the number of time periods a shift deviates from ideal shift duration (if a shift has either shorter or longer duration than ideal). So very short or long shifts are penalized harder than shifts that only have a small deviation. The weight is currently applied for both excess and deficit durations, but could (and perhaps should) be separated into two weights, differentiating between too short shifts and too long shifts.

File: utils/weights.py (hunk 7 above)
Comment (ekallevik) [reply utils/weights.py:105]: If my understanding is correct:
1. The user sets `weight_contracted_hours` to 4 demand deviation / hour 
2. You have to scale this to make sure that the objective is not affected by the size of `time_step`. 
3. You therefore multiple `weight_contracted_hours` with `time_step`, resulting in: 

```python
weight_contracted_hours = 4 demand deviation / hour * 0.25 hour = 1 demand devation per time_step
```

So you are scaling _from_ hours, and _towards_ `time_step`. If that is the case a better name would be `scale_to_time_step`. (As you are within the file `weights.py` it is not strictly necessary to include _weights_ in your function names)
